### ISSUE TITLE
Why multi same type will be set to type section after invoking `TypeBuilderSetStructType` several times?

### ISSUE THREAD

**Issue by yviansu** [WebAssembly/binaryen#5369]
I need to use customed struct type, so I used the `TypeBuilderCreate` function, `TypeBuilderSetStructType` function and `TypeBuilderBuildAndDispose` function to wrap a `createStructType` function. Just like the example below:
https://github.com/WebAssembly/binaryen/blob/94a45c6aba605f0f7e0a2fac227a2dd7c03a391f/test/example/c-api-kitchen-sink.c#L511-L529

For multi objects which has same struct structure, I may call the `createStructType` function several times, which will create the same type. However, these same types will be reflected in the type section of the generated `wat` file (like `(type ${mut:ref?|{}|_mut:f64} (struct (field (mut (ref null ${}))) (field (mut f64))))`): 
```
(module
 (type ${mut:ref?|{}|_mut:f64} (struct (field (mut (ref null ${}))) (field (mut f64))))
 (type ${mut:ref?|{}|_mut:f64} (struct (field (mut (ref null ${}))) (field (mut f64))))
 (type $none_=>_none (func))
 (type ${mut:ref?|{}|_mut:f64} (struct (field (mut (ref null ${}))) (field (mut f64))))
 (func $objTest
  (block
   xxx
  )
 )
)
```
When I use `wasm-as` to convert the `wat` file to `wasm` file, an error will be reported `[parse exception: duplicate function type (at 3:1)]Fatal: error in parsing input`.

Is there any way to avoid generating the same type in the type section when multiple calls to `createStructType` are required?

**Comment by tlively** [comment]
Are you using the nominal type system? That would cause different versions of the same type to appear separately in the type section. You may want to use the standard isorecursive type system instead, especially since we hope to remove the nominal type system soon.

The parse exception you're getting is due to a limitation of our wat printer: it generates the same name for types with the same contents. You should be able to work around the bug by emitting your module directly as a binary instead of as text, then converting the binary to text as necessary.

**Comment by yviansu** [comment]
> Are you using the nominal type system?
Yes, I didn't explicitly specify the type system, but I guess the version of `binaryen` I'm using (version 110) defaults to the nominal type system.

I have to generate wat code first because I need to use gc features, which means `-all` flag should be used in the `wasm-as` command. Can I emit the module directly as a binary with the `-all` flag is needed？

**Comment by tlively** [comment]
The features shouldn’t make any difference for emitting to wat or binary. I don’t quite understand the potential problem you’re worried about, but if there is a failing command you would expect to work, I can take a look.

**Comment by yviansu** [comment]
Thanks for your reply!

I used the `struct` related opcode through function `BinaryenStructNew` and other functions:
1. If I emit the module directly as text, I can get `xx.wat` file, and get its corresponding wasm file through command `wasm-as xx.wat -all -o xx.wasm`. 
2. If I emit the module directly as binary, I can get `yy.wasm` file, however, if I use the command `wasm-as yy.wasm -all -o yy.wat` to generate its corresponding wat file, an error `[parse exception: expected more elements in list]Fatal: error in parsing input` will occur. If I use the command `wasm2wat yy.wasm -o yy.wat`, an error `000000c: error: unexpected type form (got -0x21)` will occur. So, I can not get the wat file.

I can't get the corresponding `yy.wat` if I emit binaryen first. Is the command `wasm-as yy.wasm -all -o yy.wat` I am using wrong?

**Comment by tlively** [comment]
Aha, to go from binary .wasm to text .wat you need to use `wasm-dis` (for “disassemble”), not `wasm-as` (for “assemble”).

**Comment by yviansu** [comment]
Got it! `wasm-dis` works well for most cases.

> I used the struct related opcode through function BinaryenStructNew and other functions:
> 
> If I emit the module directly as text, I can get xx.wat file, and get its corresponding wasm file through command wasm-as xx.wat -all -o xx.wasm.
> If I emit the module directly as binary, I can get yy.wasm file, however, if I use the command wasm-as yy.wasm -all -o yy.wat to generate its corresponding wat file, an error [parse exception: expected more elements in list]Fatal: error in parsing input will occur. If I use the command wasm2wat yy.wasm -o yy.wat, an error 000000c: error: unexpected type form (got -0x21) will occur. So, I can not get the wat file.

For `xx.wasm`, `wasm-dis` can transform `xx.wasm` to `xxx.wat` successfully. However, for `yy.wasm`, when I run `./wasm-dis yy.wasm -o yy.wat`, an error `[parse exception: Expected 0 or 1 for mutability] Fatal: error in parsing wasm binary` occurs.

I think this is because the wasm I want to generate used the gc related opcode, so if I just simply get the binary by `emitBinary`, something wrong will happen, am I right?

**Comment by tlively** [comment]
GC support is very stable now, so the issue here should be fixed. Please open a new issue if there are any remaining problems.
